src/wallet/rpcdump.cpp

```diff
@@ -923,154 +924,256 @@ static std::string RecurseImportData(const CScript& script, ImportData& import_d
-                    }
-                }
-            }
-
-            if (!error.empty()) {
-                warnings.push_back("Importing as non-solvable: " + error + ". If this is intentional, don't provide any keys, pubkeys, witnessscript, or redeemscript.");
-                import_data = ImportData();
-                pubkey_map.clear();
-                privkey_map.clear();
-                have_solving_data = false;
-            } else {
-                // RecurseImportData() removes any relevant redeemscript/witnessscript from import_data, so we can use that to discover if a superfluous one was provided.
-                if (import_data.redeemscript) warnings.push_back("Ignoring redeemscript as this is not a P2SH script.");
-                if (import_data.witnessscript) warnings.push_back("Ignoring witnessscript as this is not a (P2SH-)P2WSH script.");
-                for (auto it = privkey_map.begin(); it != privkey_map.end(); ) {
-                    auto oldit = it++;
-                    if (import_data.used_keys.count(oldit->first) == 0) {
-                        warnings.push_back("Ignoring irrelevant private key.");
-                        privkey_map.erase(oldit);
-                    }
-                }
-                for (auto it = pubkey_map.begin(); it != pubkey_map.end(); ) {
-                    auto oldit = it++;
-                    auto key_data_it = import_data.used_keys.find(oldit->first);
-                    if (key_data_it == import_data.used_keys.end() || !key_data_it->second) {
-                        warnings.push_back("Ignoring public key \"" + HexStr(oldit->first) + "\" as it doesn't appear inside P2PKH or P2WPKH.");
-                        pubkey_map.erase(oldit);
-                    }
-                }
-            }
+        ImportData import_data;
+        std::map<CKeyID, CPubKey> pubkey_map;
+        std::map<CKeyID, CKey> privkey_map;
+        std::set<CScript> script_pub_keys;
+        bool have_solving_data;
+
+        if (data.exists("scriptPubKey") && data.exists("desc")) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "Both a descriptor and a scriptPubKey should not be provided.");
+        } else if (data.exists("scriptPubKey")) {
+            warnings = ProcessImportLegacy(import_data, pubkey_map, privkey_map, script_pub_keys, have_solving_data, data);
+        } else if (data.exists("desc")) {
+            warnings = ProcessImportDescriptor(import_data, pubkey_map, privkey_map, script_pub_keys, have_solving_data, data);
+        } else {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "Either a descriptor or scriptPubKey must be provided.");
         }
 
         // Check whether we have any work to do
-        if (::IsMine(*pwallet, script) & ISMINE_SPENDABLE) {
-            throw JSONRPCError(RPC_WALLET_ERROR, "The wallet already contains the private key for this address or script");
+        for (const CScript& script : script_pub_keys) {
```

Comment:
I could imagine this being an annoying failure mode if you try to import a ranged descriptor where the range is over something you've already imported. The error message doesn't give any indication of which script_pub_key caused the import to fail.

Reply:
Fixed